### DISCUSSION TITLE
Correct way to do functional tests with Jest?

### DISCUSSION THREAD

**Discussion by sebj54** [directus/directus#10283] [Archive]
Hi Directus folks :)

I am actually adding functional tests to a Directus instance. The goal is to test if our roles and permissions are correct (permissions are complex in this project).

For instance, I would like to test if an anonymous user can list items from a specific collection. I will use the JS SDK inside the test to perform the request.

But first, I need to create a record before the test. I thought that I could use `ItemsService` for that. If I'm not wrong, I can play with `accountability` to be considered as an admin.

So here is what I do in my test file:

```js
const { ItemsService } = require('directus')

test('dummy test', async () => {
    const newsService = new ItemsService('news')
    const result = await newsService.createOne({ /* … */ })
    console.log(result)

    expect(true).toBeTruthy()
})
```

The problem is it throws this error:

```
> jest

 FAIL  tests/news.test.js
  ● Test suite failed to run

    Cannot find module '@directus/shared/exceptions' from 'node_modules/directus/dist/exceptions/forbidden.js'

    Require stack:
      node_modules/directus/dist/exceptions/forbidden.js
      node_modules/directus/dist/exceptions/index.js
      node_modules/directus/dist/controllers/activity.js
      node_modules/directus/dist/app.js
      node_modules/directus/dist/index.js
      tests/news.test.js

      at Resolver.resolveModule (node_modules/jest-resolve/build/resolver.js:324:11)
      at Object.<anonymous> (node_modules/directus/dist/exceptions/forbidden.js:4:22)
```

This makes me thing I'm doing something wrong 🤷‍♂️ Is `ItemsService` intended to be used like that (I mean, outside of Directus)? 

Maybe it is related to Babel? I read Jest docs and Babel seems to be enabled by default. By the way, I installed it with:

```
npm i -D babel-jest @babel/core @babel/preset-env
```

And I also created a `babel.config.js`  file:

```js
module.exports = {
    presets: [
        [
            '@babel/preset-env',
            {
                targets: {
                    node: 'current',
                },
            },
        ],
    ],
}
```

My Jest config is the default except `setupFiles: ['dotenv/config']`.

Thanks in advance for your help! 🙏

**Comment by rijkvanzanten** [comment]
Heya!

Great question. I haven't had the need, nor have seen anybody try to use the Directus internals for custom tests outside of Directus itself, so take whatever i'm going to say next with a grain of salt.

One thing I noticed is that this is missing the context object in the parameter:

```js
    const newsService = new ItemsService('news')
```

That should be 

```js
    const newsService = new ItemsService('news', { schema });
```

where `schema` can be created with `getSchema`, which I believe should be available from `directus/utils/get-schema`.


> Is ItemsService intended to be used like that (I mean, outside of Directus)?

Not really! Though that shouldn't prevent us from trying this 🙂 

That error you're seeing is kinda odd. `@directus/shared` is a dependency from `directus`, so if you have the `directus` package installed, I'd expect it to have `@directus/shared` available as well. Maybe Jest isn't playing nice with the export/import structure between the files? Not sure!

**Reply by sebj54** [reply]
Thanks for your quick reply! Yes the error is odd and I also think it is due to the `export` used in `@directus/shared`. The `directus` package uses `require`, so that's why I tried to setup Babel. 

Do you have any tips on this? Maybe I missed something in the config, because when I try to explicitly transpile `@directus/shared`, it does not change anything.

Once I'll get through the error, I'll try to use ItemsService and will let you know if it worked. Thanks for the headsup about the constructor 😃

**Reply by rijkvanzanten** [reply]
> Yes the error is odd and I also think it is due to the export used in @directus/shared. The directus package uses require, so that's why I tried to setup Babel.

I think you've found yourself in the nightmare that is CJS (`require`) + ESM (`import`) + Jest interoperability.. I believe we ship our packages in multiple formats, that are served based on the use of the "exports" flag in package.json. It sounds to me like Jest might be pulling in the wrong version down the tree 🙃 @joselcvarela @nickrum any thoughts?

**Reply by sebj54** [reply]
That's what I'm afraid of :sob:

Maybe the simplest fix is to use something else than Jest?

**Reply by rijkvanzanten** [reply]
> Maybe the simplest fix is to use something else than Jest?

I'm not sure.. though it's worth a shot! @jaycammarano didn't we run into something similar on the Jest side?

**Reply by jamescammarano** [reply]
I didn't but I think @joselcvarela did.

**Reply by joselcvarela** [reply]
Hello all,
Really grateful to you for bringing me into this again 😅
@sebj54 it is a nightmare setting up Jest and use import/export (ESM). Maybe you can find something useful here https://github.com/directus/directus/pull/8524

Right now my advice is to change your `require` into a dynamic import, like so:
```js
test('dummy test', async () => {
    const { ItemsService } = await import('directus')
    const newsService = new ItemsService('news')
    const result = await newsService.createOne({ /* … */ })
    console.log(result)

    expect(true).toBeTruthy()
})
```

**Comment by sebj54** [comment]
Thank you for all your comments, I finally figured it out thanks to you!

There were two problems:

- Jest does not support `package.json`'s `exports` field and that's what `@directus/shared` uses. 
See: https://github.com/directus/directus/blob/main/packages/shared/package.json and https://github.com/facebook/jest/issues/9771
- Jest does not support import/export (ESM) **for dependencies** out of the box

Here is how I fixed it:

- I now use [`jest-node-exports-resolver`](https://github.com/k-g-a/jest-node-exports-resolver) as a resolver with Jest
- I explicitly transpile `@directus/shared` with Babel for ESM support

**jest.config.js**

```js
module.exports = {
    resolver: 'jest-node-exports-resolver',
    transformIgnorePatterns: [
        '/node_modules/(?!@directus/shared)',
    ],
}
```

And now, I can use `ItemsService` in my tests like I wanted (and it does what I needed):

**news.test.js**

```js
import { ItemsService } from 'directus'
import { getSchema } from 'directus/utils/get-schema'

test('dummy test', async () => {
    const schema = await getSchema()
    const newsService = new ItemsService('news', { schema })
    const result = await newsService.createOne({ /* … */ })
    console.log(result)

    expect(true).toBeTruthy()
})
```

**Reply by keesvanbemmel** [reply]
@sebj54 thanks for your insights! Would you be able to share some tests you've written? I'm particularly looking for ways to mock contexts and ItemsService calls for example...

**Reply by sebj54** [reply]
@keesvanbemmel Unfortunately I didn't mock anything, I only wrote functional tests

**Comment by derfriedrich** [comment]
In case someone finds this discussion on how to write functional/integration tests for directus, here is my currently working setup (directus 11.1) with "node:test" (available since Node.js v20). 

**test/directus.js**
```js
import { describe, test, after } from "node:test";
import assert from "node:assert";

import { ItemsService } from "@directus/api/services/index";
import { getSchema } from "@directus/api/utils/get-schema";
import { getDatabase } from "@directus/api/database/index";

const database = getDatabase();
const schema = await getSchema();
const userService = new ItemsService("directus_users", { schema });

describe("directus test", () => {
  test("can read users", async (t) => {
    const users = await userService.readByQuery({
      fields: ["*"],
    });
    assert.equal(users.length > 0, true);
  });
  after(async () => {
    // always close the database connection after all tests are done
    await database.destroy();
  });
});
```

Run the test with:
```sh
node --test test/directus.js
```

**Comment by NilsBaumgartner1994** [comment]
> In case someone finds this discussion on how to write functional/integration tests for directus, here is my currently working setup (directus 11.1) with "node:test" (available since Node.js v20). 
> 
> 
> 
> **test/directus.js**
> 
> ```js
> 
> import { describe, test, after } from "node:test";
> 
> import assert from "node:assert";
> 
> 
> 
> import { ItemsService } from "@directus/api/services/index";
> 
> import { getSchema } from "@directus/api/utils/get-schema";
> 
> import { getDatabase } from "@directus/api/database/index";
> 
> 
> 
> const database = getDatabase();
> 
> const schema = await getSchema();
> 
> const userService = new ItemsService("directus_users", { schema });
> 
> 
> 
> describe("directus test", () => {
> 
>   test("can read users", async (t) => {
> 
>     const users = await userService.readByQuery({
> 
>       fields: ["*"],
> 
>     });
> 
>     assert.equal(users.length > 0, true);
> 
>   });
> 
>   after(async () => {
> 
>     // always close the database connection after all tests are done
> 
>     await database.destroy();
> 
>   });
> 
> });
> 
> ```
> 
> 
> 
> Run the test with:
> 
> ```sh
> 
> node --test test/directus.js
> 
> ```

Would the database needs to exist in first place? In your example the test should fail, since no user was created before?

**Comment by NilsBaumgartner1994** [comment]
Related to https://github.com/directus/directus/discussions/24692
